src/sources/dnstap/parser.rs

```diff
@@ -0,0 +1,1053 @@
+use crate::{
+    event::{LogEvent, PathComponent, Value},
+    Error, Result,
+};
+use bytes::Bytes;
+use prost::Message;
+use snafu::Snafu;
+use std::net::{IpAddr, Ipv4Addr, Ipv6Addr};
+use std::{convert::TryInto, fmt::Debug};
+use trust_dns_proto::{
+    rr::domain::Name,
+    serialize::binary::{BinDecodable, BinDecoder},
+};
+mod dnstap_proto {
+    include!(concat!(env!("OUT_DIR"), "/dnstap.rs"));
+}
+
+use dnstap_proto::{
+    dnstap::Type as DnstapDataType, message::Type as DnstapMessageType, Dnstap,
+    Message as DnstapMessage, SocketFamily, SocketProtocol,
+};
+
+use super::dns_message::{
+    DnsRecord, EdnsOptionEntry, OptPseudoSection, QueryHeader, QueryQuestion, UpdateHeader,
+    ZoneInfo,
+};
+use super::dns_message_parser::DnsMessageParser;
+use super::schema::DnstapEventSchema;
+
+#[derive(Debug, Snafu)]
+enum DnstapParserError {
+    #[snafu(display("Unsupported DNSTap message type: {}", "dnstap_message_type_id"))]
+    UnsupportedDnstapMessageTypeError { dnstap_message_type_id: i32 },
+}
+
+pub struct DnstapParser<'a> {
+    event_schema: &'a DnstapEventSchema,
+    parent_key_path: Vec<PathComponent>,
+    log_event: &'a mut LogEvent,
+}
+
+impl<'a> DnstapParser<'a> {
+    pub fn new(event_schema: &'a DnstapEventSchema, log_event: &'a mut LogEvent) -> Self {
```

Comment:
In fact that's exactly how the parser was originally planned to implement. However, to some point, we saw the monolithic function grew so big that it's basically impossible to maintain and test anymore. So we decided to split it into multiple smaller but maintainable functions, and realized that there were a couple of variables (e.g. the schema, `log_event`, etc.) that had to be passed along with each call of these functions. That made the function signatures bloated. Eventually, we came up with the current parser design.

Reply:
Makes sense. Do you think we could still encapsulate that in this module? That is just have one function that takes the event schema and bytes, and returns a log event? It could still pass around a mutable log event in here that was simpler.

Reply:
It has been encapsulated in a function, which is called here instead.